### PR TITLE
refactor: Update median tracking for historacle pricing

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #1621

Things that still need to be done after this PR:
- Add new median stat keeper methods as grpc queries
- Build out oracle endblocker integration test to test historacle functionality

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/keeper/historic_price.go [hunk 1]
```diff
@@ -47,55 +47,91 @@ func medianDeviation(median sdk.Dec, prices []sdk.Dec) sdk.Dec {
 	return medianDeviation
 }
 
-// GetMedian returns a given denom's median price in the last prune
-// period since a given block.
-func (k Keeper) GetMedian(
+// average returns the average value of a list of prices
```
Comment (zarazan) [inline x/oracle/keeper/historic_price.go:50]: What do you think about moving all these functions that don't interact with the keeper somewhere else? They are mostly just calculations done on a set of prices, maybe something like `price_calculations.go`

File: x/oracle/keeper/migrations.go [hunk 2]
```diff
@@ -18,9 +16,9 @@ func NewMigrator(keeper *Keeper) Migrator {
 
 // Migrate1to2 migrates from version 1 to 2.
 func (m Migrator) Migrate1to2(ctx sdk.Context) error {
-	m.keeper.SetStampPeriod(ctx, 1)
-	m.keeper.SetPrunePeriod(ctx, 1)
-	m.keeper.SetMedianPeriod(ctx, 1)
-	m.keeper.SetHistoricAcceptList(ctx, types.DenomList{})
+	m.keeper.SetHistoricStampPeriod(ctx, 1)
```
Comment (zarazan) [inline x/oracle/keeper/migrations.go:19]: Does it matter that we are not removing the old values?

File: x/oracle/keeper/historic_price.go [hunk 3]
```diff
@@ -163,6 +253,25 @@ func (k Keeper) getHistoricPrices(
...
+// range.
+func (k Keeper) getMedians(
+	ctx sdk.Context,
+	denom string,
+	startBlock uint64,
+	endBlock uint64,
+) []sdk.Dec {
+	medians := []sdk.Dec{}
+	k.IterateMedians(ctx, denom, func(exchangeRate sdk.Dec, block uint64) bool {
+		if block >= startBlock && block <= endBlock {
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:266]: `startBlock` should be included in the key prefix search, to avoid traversing and deserializing not needed records.

File: x/oracle/keeper/historic_price.go [hunk 4]
```diff
@@ -187,6 +296,30 @@ func (k Keeper) IterateHistoricPrices(
...
+// Iterator stops when exhausting the source, or when the handler returns `true`.
+func (k Keeper) IterateMedians(
+	ctx sdk.Context,
+	denom string,
+	handler func(sdk.Dec, uint64) bool,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	// make sure we have one zero byte to correctly separate denoms
+	prefix := util.ConcatBytes(1, types.KeyPrefixMedian, []byte(denom))
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:310]: we need to have separator. Otherwise the code is not valid. Consider 2 denoms: `eth`, ethpow`. When searching for `eth` one, `ethpow` will be returned as well.

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
Need to fix `IterateMedians`

File: x/oracle/keeper/historic_price.go (hunk 1 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: Added them in our utils package in pricestats.go 👍

File: x/oracle/keeper/historic_price.go (hunk 3 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: This is tricky with the KVStore iterator since the prefix is changing every iteration. I changed it to not use the iterator and accomplish the same result in `getMedians()`. Let me know what you think about that solution.

File: x/oracle/keeper/historic_price.go (hunk 4 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: I believe this already taken care of by the util methods `ConcatBytes()` and `UintWithNullPrefix()`. I updated `historic_price_test.go` to test for this and it looks like this edge case is covered for.

File: x/oracle/keeper/migrations.go (hunk 2 above)
Comment (rbajollari) [reply x/oracle/keeper/migrations.go:19]: I'm not actually sure. @robert-zaremba would this be an issue?

**Review by zarazan — COMMENTED** [review]
Great job getting this together! It needed a ton of changes since last time. I think we're almost there, a couple tiny comments from me and some consensus on the iteration method and I think its good.

File: x/oracle/keeper/historic_price.go [hunk 5]
```diff
@@ -1,261 +1,302 @@
...
+        k.cdc.MustUnmarshal(iter.Value(), &decProto)
+        if handler(decProto.Dec) {
+            break
+        }
+    }
+}
+
+// getMedians returns median prices of a given denom in the store
+// starting at startblock until the latest block.
+func (k Keeper) getMedians(
```
[the remark this answers is not in the document]
Comment (zarazan) [reply x/oracle/keeper/historic_price.go]: How does this resolve the issue Robert mentioned here: https://github.com/umee-network/umee/pull/1632#discussion_r1037723821 ? Is a separator not needed?

[the remark this answers is not in the document]
Comment (zarazan) [reply x/oracle/keeper/historic_price.go]: Oh yeah, that will work. Does this match the key? The margin passed to `ConcatBytes` is 0. https://github.com/umee-network/umee/blob/a0630a097a82dfa3d5a4c5a1dd9f36efcca3bef5/x/oracle/types/keys.go#L65

[the remark this answers is not in the document]
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: Yeah I believe the UintWithNullPrefix util takes care of the null byte after the denom there. I also test for this in `historic_price_test.go` by adding prices and medians to a denom with a similar prefix ("UMEEtest" along with "UMEE") to check if the "UMEEtest" prices/medians being tracked mess with the calculations for "UMEE".

**Review by zarazan — APPROVED** [review]
I think @robert-zaremba's concerns are resolved but idk if he wants to take a look before merging.

File: x/oracle/keeper/historic_price.go (hunk 4 above)
Comment (robert-zaremba) [reply x/oracle/keeper/historic_price.go:248]: yes, now `ConcatBytes` solves this.

File: util/pricestats.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+func Min(prices []sdk.Dec) (sdk.Dec, error) {
+	lenPrices := len(prices)
+	if lenPrices == 0 {
+		return sdk.ZeroDec(), ErrEmptyList
+	}
+
+	sort.Slice(prices, func(i, j int) bool {
+		return prices[i].BigInt().
+			Cmp(prices[j].BigInt()) < 0
+	})
```
Comment (robert-zaremba) [inline util/pricestats.go:97]: don't use sort to find Min

File: util/pricestats.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+func Max(prices []sdk.Dec) (sdk.Dec, error) {
+	lenPrices := len(prices)
+	if lenPrices == 0 {
+		return sdk.ZeroDec(), ErrEmptyList
+	}
+
+	sort.Slice(prices, func(i, j int) bool {
+		return prices[i].BigInt().
+			Cmp(prices[j].BigInt()) < 0
+	})
```
Comment (robert-zaremba) [inline util/pricestats.go:81]: don't use sort to find `Max`

File: util/pricestats.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
+package util
```
Comment (robert-zaremba) [inline util/pricestats.go:1]: let's move this file to a more specific package, eg : `util/decmath`

File: x/oracle/abci.go [hunk 9]
```diff
@@ -63,14 +59,16 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper, experimental bool) error {
 			}
 
 			if experimental {
-				// Stamp rate every stamp period if asset is set to have historic stats tracked
-				if isPeriodLastBlock(ctx, params.StampPeriod) && params.HistoricAcceptList.Contains(ballotDenom.Denom) {
-					k.AddHistoricPrice(ctx, ballotDenom.Denom, exchangeRate)
+				// Stamp historic price if historic stamp period has passed
+				if isPeriodLastBlock(ctx, params.HistoricStampPeriod) {
+					k.AddHistoricPrice(ctx, strings.ToUpper(ballotDenom.Denom), exchangeRate)
```
Comment (robert-zaremba) [inline x/oracle/abci.go:64]: shouldn't this be normalized when the ballot is created?

File: util/pricestats_test.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	require.ErrorIs(err, ErrEmptyList)
+}
+
+func TestAverage(t *testing.T) {
+	require := require.New(t)
+	prices := []sdk.Dec{
+		sdk.MustNewDecFromStr("1.1"),
+		sdk.MustNewDecFromStr("1.05"),
+		sdk.MustNewDecFromStr("1.15"),
+		sdk.MustNewDecFromStr("1.2"),
```
Comment (robert-zaremba) [inline util/pricestats_test.go:53]: let's use different data, such that avg != median.

File: x/oracle/keeper/historic_price.go [hunk 11]
```diff
@@ -11,91 +10,65 @@ import (
...
+func (k Keeper) GetHistoricMedians(
 	ctx sdk.Context,
 	denom string,
-) (sdk.Dec, error) {
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.KeyMedian(denom))
-	if bz == nil {
-		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrNoMedian, fmt.Sprintf("denom: %s", denom))
-	}
+	numStamps uint64,
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:17]: :+1:

File: x/oracle/keeper/migrations.go (hunk 2 above)
Comment (robert-zaremba) [reply x/oracle/keeper/migrations.go:19]: In migration? But on mainnet, nor canon-2 the medians and historic prices won't be recorded. So there is nothing to remove.
The historic prices should only be recorded (now) when `experimental == true`).

**Review by robert-zaremba — COMMENTED** [review]
thanks for fixing the denom prefix for iterations.

- [ ] Please verify if these change are in line with the updated: https://github.com/umee-network/umee/commit/3909192e39288bdeccfd6a634cfb2ccac2463c32

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
And let's remove the new files from the /util package

File: x/oracle/abci.go (hunk 9 above)
Comment (rbajollari) [reply x/oracle/abci.go:64]: If `VotePeriod` == `HistoricStampPeriod` yes, but we decided on having `HistoricStampPeriod` be a parameter to be able to test whether or not prices need to be stamped as frequent as the `VotePeriod`.

File: util/pricestats.go (hunk 6 above)
Comment (rbajollari) [reply util/pricestats.go]: Updated 👍

File: util/pricestats.go (hunk 7 above)
Comment (rbajollari) [reply util/pricestats.go]: Updated 👍

**Comment by rbajollari** [comment]
@robert-zaremba @zarazan @adamewozniak Updated the keeper methods in the historacle design doc to match the actual implementation in this PR (some of the stuff agreed upon in #1631 wasn't put into the doc). Lmk if anything looks wrong in there.

File: docs/design_docs/011-historacle-pricing.md [hunk 12]
```diff
@@ -63,12 +63,12 @@ We define two Maximum values, which correspond to the most we will store of a me
...
-- `MaxMedian(denom string, blockNum int) sdk.Dec` returns the Maximum of all the Medians recorded within the past `blockNum`.
-- `MinMedian(denom string, blockNum int) sdk.Dec` returns the Minimum of all the Medians recorded within the past `blockNum`.
+- `GetHistoricMedians(denom string, numStamps uint64) []sdk.Dec` returns list of last `numStamps` amount of median prices of an asset
+- `WithinHistoricDeviation(denom string) (bool, error)` returns whether or not the current price of an asset is within the Standard Deviation around the Median.
+- `GetMedianOfMedians(denom string, numStamps uint64) (sdk.Dec, error)` returns the Median of the all the Medians recorded within the past `numStamps` of medians.
+- `GetAverageOfMedians(denom string, numStamps uint64) (sdk.Dec, error)` returns the Average of all the Medians recorded within the past `numStamps` of medians.
+- `GetMaxOfMedians(denom string, numStamps uint64) (sdk.Dec, error)` returns the Maximum of all the Medians recorded within the past `numStamps` of medians.
+- `GetMinOfMedians(denom string, numStamps uint64) (sdk.Dec, error)` returns the Minimum of all the Medians recorded within the past `numStamps` of medians.
```
Comment (robert-zaremba) [inline docs/design_docs/011-historacle-pricing.md:71]: This is not Java. Why prefixing every function with `Get`?

Comment (rbajollari) [reply docs/design_docs/011-historacle-pricing.md:71]: This was to be consistent with other keeper methods e.g. `GetAggregateExchangeRatePrevote` and to avoid query functions and keeper functions sharing the same name like `MedianOfMedians`. I agree with not using `Get` though, so I've just added the `historic` key word to those method names instead of using `Get` to address the second issue.

Comment (robert-zaremba) [reply docs/design_docs/011-historacle-pricing.md:71]: Many of these methods do computation (eg, Avg of medians), rather than just fetching a data. 
Following a bad practice is a bad practice ;) 

Anyway, this is not a blocking comment.

File: x/oracle/abci.go [hunk 13]
```diff
@@ -108,11 +106,15 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper, experimental bool) error {
...
-	// Prune historic prices every prune period
-	if isPeriodLastBlock(ctx, params.PrunePeriod) && experimental {
-		pruneBlock := uint64(ctx.BlockHeight()) - params.PrunePeriod
-		for _, v := range params.HistoricAcceptList {
-			k.DeleteHistoricPrice(ctx, v.String(), pruneBlock)
+	// Prune historic prices and medians outside pruning period determined by
+	// the stamp period multiplied by the max stamps.
+	if experimental {
+		pruneHistoricBlock := uint64(ctx.BlockHeight()) - (params.HistoricStampPeriod * params.MaximumPriceStamps)
+		pruneMedianBlock := uint64(ctx.BlockHeight()) - (params.MedianStampPeriod * params.MaximumMedianStamps)
```
Comment (robert-zaremba) [inline x/oracle/abci.go:113]: how about avoiding prunning every block, and instead batch it, eg once every 100 blocks (10min).

File: x/oracle/keeper/genesis.go [hunk 14]
```diff
@@ -43,8 +43,12 @@ func (k Keeper) IterateAllMedianPrices(
 	for ; iter.Valid(); iter.Next() {
 		var decProto sdk.DecProto
 		k.cdc.MustUnmarshal(iter.Value(), &decProto)
-		denom := types.ParseDenomFromMedianKey(iter.Key())
-		median := types.ExchangeRateTuple{ExchangeRate: decProto.Dec, Denom: denom}
+		denom := types.ParseDenomFromKey(iter.Key(), types.KeyPrefixMedian)
+		blockNum := types.ParseBlockFromKey(iter.Key())
```
Comment (robert-zaremba) [inline x/oracle/keeper/genesis.go:47]: can we have one function which will return both denom and a block number?

File: x/oracle/keeper/grpc_query_test.go [hunk 15]
```diff
@@ -203,33 +203,39 @@ func (s *IntegrationTestSuite) TestQuerier_AggregateVotesAppendVotes() {
...
 
 func (s *IntegrationTestSuite) TestQuerier_Medians() {
+	app, ctx := s.app, s.ctx
+
 	atomMedian := sdk.DecCoin{Denom: "atom", Amount: sdk.MustNewDecFromStr("49.99")}
 	umeeMedian := sdk.DecCoin{Denom: "umee", Amount: sdk.MustNewDecFromStr("6541.48")}
 
-	s.app.OracleKeeper.SetMedian(s.ctx, atomMedian.Denom, atomMedian.Amount)
-	s.app.OracleKeeper.SetMedian(s.ctx, umeeMedian.Denom, umeeMedian.Amount)
+	app.OracleKeeper.SetMedianStampPeriod(ctx, 1)
```
Comment (robert-zaremba) [inline x/oracle/keeper/grpc_query_test.go:211]: Let's have a test with period != 1

File: x/oracle/keeper/historic_price.go [hunk 16]
```diff
@@ -106,50 +80,126 @@ func (k Keeper) GetMedianDeviation(
...
+		return false, sdkerrors.Wrap(types.ErrNoMedian, fmt.Sprintf("denom: %s", denom))
+	}
+	median := medians[0]
+
+	// get latest historic price
+	prices := k.historicPrices(ctx, denom)
+	if len(prices) == 0 {
+		return false, sdkerrors.Wrap(types.ErrNoHistoricPrice, fmt.Sprintf("denom: %s", denom))
 	}
+	price := prices[0]
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:102]: if we want only the latest price, why do we deserialize all of them? Please update, (you can use reverse iterator) to only fetch max one price. (it's possible that there will be zero)

File: x/oracle/keeper/historic_price.go [hunk 17]
```diff
@@ -187,6 +237,33 @@ func (k Keeper) IterateHistoricPrices(
...
+	iter := sdk.KVStoreReversePrefixIterator(store, prefix)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		decProto := sdk.DecProto{}
+		k.cdc.MustUnmarshal(iter.Value(), &decProto)
+		block := types.ParseBlockFromKey(iter.Key())
+		if handler(decProto.Dec) || block <= startBlock {
+			break
+		}
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:263]: Let's add start block to the iterator prefix. No need to access DB records we don't need.

File: x/oracle/keeper/historic_price.go (hunk 16 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go:102]: Update to iterate until a start block like the medians iterator 👍

File: x/oracle/keeper/grpc_query_test.go (hunk 15 above)
Comment (rbajollari) [reply x/oracle/keeper/grpc_query_test.go:211]: I'll do so in the next pr for updating the grpc queries

File: x/oracle/keeper/genesis.go (hunk 14 above)
Comment (rbajollari) [reply x/oracle/keeper/genesis.go:47]: Done👍

File: x/oracle/abci.go (hunk 13 above)
Comment (rbajollari) [reply x/oracle/abci.go:112]: I changed it to prune every historic stamp period. Any less frequent would result in us storing more than the maximum price stamps.

File: util/decmath/decmath.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+	for _, price := range prices {
+		sumPrices = sumPrices.Add(price)
+	}
+
+	return sumPrices.QuoInt64(int64(lenPrices)), nil
+}
+
+// Max returns the max value of a list of prices. Returns error
+// if prices is empty list.
+func Max(prices []sdk.Dec) (sdk.Dec, error) {
```
[the remark this answers is not in the document]
Comment (rbajollari) [reply util/decmath/decmath.go:71]: Done👍

File: x/oracle/abci.go (hunk 9 above)
Comment (rbajollari) [reply x/oracle/abci.go]: Initially yes but I changed it back to not convert to upper case to keep consistent with the `SetExchangeRate` method in the endblocker

File: x/oracle/keeper/historic_price.go [hunk 19]
```diff
@@ -106,50 +80,126 @@ func (k Keeper) GetMedianDeviation(
...
+	denom string,
+	numStamps uint64,
+) (sdk.Dec, error) {
+	medians := k.HistoricMedians(ctx, denom, numStamps)
+	max, err := decmath.Max(medians)
+	if err != nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(err, fmt.Sprintf("denom: %s", denom))
+	}
+	return max, nil
+}
```
[the remark this answers is not in the document]
Comment (zarazan) [reply x/oracle/keeper/historic_price.go:184]: https://github.com/umee-network/umee/pull/1631 Let's leave them in for now since that is what we already decided over here: https://github.com/umee-network/umee/pull/1631

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/oracle/keeper/historic_price.go:184]: It should be avoided. User should not call `MedianOfMedian` or `MaxMedian` through the oracle keeper, because it will cause to fetch all medians twice. So at minimum, if we want to keep it, we should clearly describe it in the function comment.

File: x/oracle/keeper/historic_price.go (hunk 17 above)
Comment (robert-zaremba) [reply x/oracle/keeper/historic_price.go:271]: No, you can iterate from a given key,

File: x/oracle/keeper/grpc_query_test.go (hunk 15 above)
Comment (robert-zaremba) [reply x/oracle/keeper/grpc_query_test.go:211]: OK, then please add todo , or a task.

File: x/oracle/keeper/historic_price.go (hunk 16 above)
Comment (robert-zaremba) [reply x/oracle/keeper/historic_price.go:102]: I don't see the update

File: x/oracle/keeper/historic_price.go (hunk 17 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: Changed it to use KVStoreReversePrefixIteratorPaginated

File: x/oracle/keeper/historic_price.go (hunk 16 above)
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go:97]: Im calling for only the latest historic stamp now rather than calling for all of them; `prices := k.historicPrices(ctx, denom, 1)` the "1" parameter being numstamps

File: x/oracle/keeper/historic_price.go [hunk 20]
```diff
@@ -164,18 +211,44 @@ func (k Keeper) getHistoricPrices(
 }
 
 // IterateHistoricPrices iterates over historic prices of a given
-// denom in the store.
+// denom in the store in reverse.
 // Iterator stops when exhausting the source, or when the handler returns `true`.
 func (k Keeper) IterateHistoricPrices(
 	ctx sdk.Context,
 	denom string,
+	numStamps uint,
```
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:219]: :+1:

**Review by robert-zaremba — APPROVED** [review]
utACK. Thanks for keeping improving the PR.

**Comment by toteki** [comment]
Note: be sure to run `make proto-all` after these PRs.

In this case, there's a little whitespace that gets changed by `make proto-format` and some changes not reflected in `swagger/` folder.
